witness/src/Dscp/Snowdrop/PublicationValidation.hs

```diff
@@ -0,0 +1,106 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Dscp.Snowdrop.PublicationValidation
+       ( validatePublication
+       , Publication (..)
+       , PublicationValidationError (..)
+       , PublicationTx (..)
+       , PublicationTxTypeId (..)
+       ) where
+
+import Data.Map as Map (lookup)
+
+import Dscp.Core.Transactions
+import Dscp.Snowdrop.AccountValidation
+import Snowdrop.Model.State.Core (HasKeyValue, PreValidator (..), StateTx (..), StateTxType (..),
+                                  Validator, mkValidator, queryOne)
+
+import Snowdrop.Util
+
+-- | Transaction type for publication.
+data PublicationTxTypeId
+    = PublicationTxTypeId
+    deriving (Eq, Ord, Show, Generic)
+
+data PublicationValidationError
+    = PublicationSignatureIsIncorrect
+    | PublicationPrevBlockIsIncorrect
+    | StorageIsCorrupted
+    | PublicationIsBroken
+
+validatePublication
+    :: forall       e pk addr id proof value ctx sig hash stTxId
```

Comment:
You don't need to parametrise it here, please remove parameters from the core types and use concrete types like `PublicKey`, `Ids`, `Values` etc. This code won't be moved to SD, it's for our project only and this parametration only makes it more complex.